File: farumasi_api/app/services/audit_service.py
from __future__ import annotations

from typing import Optional

from sqlalchemy.ext.asyncio import AsyncSession

from app.models.audit import AuditLog


class AuditService:
    def __init__(self, db: AsyncSession):
        self.db = db

    async def log(
        self,
        action: str,
        actor_user_id: Optional[str] = None,
        entity_type: Optional[str] = None,
        entity_id: Optional[str] = None,
        old_value: Optional[dict] = None,
        new_value: Optional[dict] = None,
        ip_address: Optional[str] = None,
    ) -> AuditLog:
        entry = AuditLog(
            actor_user_id=actor_user_id,
            action=action,
            entity_type=entity_type,
            entity_id=entity_id,
            old_value=old_value,
            new_value=new_value,
            ip_address=ip_address,
        )
        self.db.add(entry)
        await self.db.flush()
        return entry

File: farumasi_super_admin/src/lib/services/audit.service.ts
import api from "@/lib/api";

export interface AuditLog {
  id: string;
  actor_user_id?: string | null;
  action: string;
  entity_type?: string | null;
  entity_id?: string | null;
  old_value?: unknown;
  new_value?: unknown;
  ip_address?: string | null;
  created_at: string;
}

export interface AuditListParams {
  offset?: number;
  limit?: number;
  action?: string;
  entity_type?: string;
  entity_id?: string;
  actor_user_id?: string;
  search?: string;
  sort_by?: "created_at" | "action";
  sort_dir?: "asc" | "desc";
}

export const auditService = {
  async list(params?: AuditListParams): Promise<{ items: AuditLog[]; total: number }> {
    const { data } = await api.get<{ items: AuditLog[]; total: number; offset: number; limit: number }>(
      "/admin/audit-logs",
      { params },
    );
    return { items: data.items, total: data.total };
  },

  async getById(logId: string): Promise<AuditLog> {
    const { data } = await api.get<AuditLog>(`/admin/audit-logs/${logId}`);
    return data;
  },

  async getEntityTypes(): Promise<string[]> {
    const { data } = await api.get<string[]>("/admin/audit-logs/meta/entity-types");
    return data;
  },
};
